File: ruby/dialog/channeldialog.rb
# Created by Satoshi Nakagawa.
# You can redistribute it and/or modify it under the Ruby's license or the GPL2.

require 'dialoghelper'

class ChannelDialog < OSX::NSObject
  include OSX
  include DialogHelper
  
  attr_accessor :delegate, :prefix, :parent
  attr_reader :uid, :cid, :modal
  ib_outlet :window
  ib_mapped_outlet :nameText, :passwordText, :modeText, :topicText, :auto_joinCheck, :keywordCheck, :unreadCheck, :consoleCheck
  ib_outlet :okButton
  
  def initialize
    @prefix = 'channelDialog'
  end
  
  def config
    @c
  end
  
  def loadNib
    NSBundle.loadNibNamed_owner('ChannelDialog', self)
  end
  
  def start(config, uid, cid)
    @c = config.dup
    @uid = uid
    @cid = cid
    loadNib
    if cid < 0
      @window.setTitle("New Channel")
    else
      @nameText.setEditable(false)
      @nameText.setSelectable(false)
      @nameText.setBezeled(false)
      @nameText.setDrawsBackground(false)
      #@nameText.setEnabled(false)
    end
    load
    update
    show
  end
  
  def start_sheet(owner, config, add=false)
    @modal = true
    @c = config.dup
    loadNib
    if add
      @c.name = ''
    else
      @nameText.setEditable(false)
      @nameText.setSelectable(false)
      @nameText.setBezeled(false)
      @nameText.setDrawsBackground(false)
    end
    load
    update
    NSApp.beginSheet_modalForWindow_modalDelegate_didEndSelector_contextInfo(@window, owner, self, 'sheetDidEnd:returnCode:contextInfo:', nil)
  end
  
  def sheetDidEnd_returnCode_contextInfo(sender, code, info)
    @window.close
    @modal = false
    if code != 0
      save
      fire_event('onOk', @c)
    else
      fire_event('onCancel')
    end
  end
  
  def show
    @window.centerOfWindow(@parent) unless @window.isVisible
    @window.makeKeyAndOrderFront(self)
  end
  
  def close
    @delegate = nil
    @window.close
  end
  
  def windowWillClose(sender)
    fire_event('onClose')
  end
  
  def onOk(sender)
    if @modal
      NSApp.endSheet_returnCode(@window, 1)
    else
      save
      fire_event('onOk', @c)
      @window.close
    end
  end
  
  def onCancel(sender)
    if @modal
      NSApp.endSheet_returnCode(@window, 0)
    else
      @window.close
    end
  end
  
  def load
    load_mapped_outlets(@c)
  end
  
  def save
    save_mapped_outlets(@c)
  end
  
  def controlTextDidChange(n)
    update
  end
  
  def update
    s = @nameText.stringValue.to_s
    @okButton.setEnabled(s.channelname?)
  end
end

File: Classes/Dialogs/ChannelDialog.m
// LimeChat is copyrighted free software by Satoshi Nakagawa <psychs AT limechat DOT net>.
// You can redistribute it and/or modify it under the terms of the GPL version 2 (see the file GPL.txt).

#import "ChannelDialog.h"
#import "NSWindowHelper.h"
#import "NSStringHelper.h"
#import "DialogWindow.h"


@implementation ChannelDialog

- (id)init
{
    self = [super init];
    if (self) {
        [[NSBundle mainBundle] loadNibNamed:@"ChannelDialog" owner:self topLevelObjects:nil];
    }
    return self;
}

- (void)start
{
    [self load];
    [self update];
    [self show];
}

- (void)startSheet
{
    [self load];
    [self update];
    [[self _dialogWindow] startSheetModalForWindow:_parentWindow];
}

- (void)show
{
    if (![self.window isVisible]) {
        [self.window centerOfWindow:_parentWindow];
    }
    [self.window makeKeyAndOrderFront:nil];
}

- (void)close
{
    _delegate = nil;
    [self.window close];
}

- (void)load
{
    _nameText.stringValue = _config.name;
    _passwordText.stringValue = _config.password;
    _modeText.stringValue = _config.mode;
    _topicText.stringValue = _config.topic;

    _autoJoinCheck.state = _config.autoJoin;
    _consoleCheck.state = _config.logToConsole;
    _notifyCheck.state = _config.notify;
}

- (void)save
{
    _config.name = _nameText.stringValue;
    _config.password = _passwordText.stringValue;
    _config.mode = _modeText.stringValue;
    _config.topic = _topicText.stringValue;

    _config.autoJoin = _autoJoinCheck.state;
    _config.logToConsole = _consoleCheck.state;
    _config.notify = _notifyCheck.state;

    if (![_config.name isChannelName]) {
        _config.name = [@"#" stringByAppendingString:_config.name];
    }
}

- (void)update
{
    if (_cid < 0) {
        [self.window setTitle:@"New Channel"];
    }
    else {
        [_nameText setEditable:NO];
        [_nameText setSelectable:NO];
        [_nameText setBezeled:NO];
        [_nameText setDrawsBackground:NO];
    }

    NSString* s = _nameText.stringValue;
    [_okButton setEnabled:s.length > 0];
}

- (void)controlTextDidChange:(NSNotification*)note
{
    [self update];
}

- (DialogWindow *)_dialogWindow
{
    return (DialogWindow *)self.window;
}

#pragma mark - Actions

- (void)ok:(id)sender
{
    [self save];

    if ([_delegate respondsToSelector:@selector(channelDialogOnOK:)]) {
        [_delegate channelDialogOnOK:self];
    }

    [self cancel:nil];
}

- (void)cancel:(id)sender
{
    [[self _dialogWindow] closeWindowOrSheet];
}

#pragma mark - NSWindow Delegate

- (void)windowWillClose:(NSNotification*)note
{
    if ([_delegate respondsToSelector:@selector(channelDialogWillClose:)]) {
        [_delegate channelDialogWillClose:self];
    }
}

@end
